File: leetcode/29.py
class Solution(object):
    def divide(self, a, b):
        x, y, ans = abs(a), abs(b), 0
        sgn = 1 if a * b > 0 else -1
        while x >= y:
            now, cnt = y, 1
            while now * 2 < x:
                now <<= 1
                cnt <<= 1
            ans += cnt
            x -= now
        ans *= sgn
        ans = min(ans, 2 ** 31 - 1)
        return ans

File: leetcode/29.cpp
typedef long long LL;
class Solution {
public:
    int divide(int xx, int yy) {
        LL x = abs((LL)xx), y = abs((LL)yy);
        int sgn = (LL)xx * yy > 0 ? 1 : -1;
        LL ans = 0;
        cout << x << "XXXX" << y << endl;
        for (LL i = 31; i >= 0; --i)
            if ((y << i) <= x) {
                ans += 1ll << i;
                x -= y << i;
            }
       // cout << ans << endl;
        return min((LL)INT_MAX, ans * sgn);
    }
};